alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/FollowFlowField.kt

```diff
@@ -1,50 +1,56 @@
 package it.unibo.alchemist.model.implementations.actions
 
 import it.unibo.alchemist.model.implementations.layers.BidimensionalGaussianLayer
-import it.unibo.alchemist.model.implementations.utils.origin
+import it.unibo.alchemist.model.interfaces.Action
+import it.unibo.alchemist.model.interfaces.Environment
+import it.unibo.alchemist.model.interfaces.Layer
 import it.unibo.alchemist.model.interfaces.Molecule
+import it.unibo.alchemist.model.interfaces.Node
 import it.unibo.alchemist.model.interfaces.Pedestrian2D
+import it.unibo.alchemist.model.interfaces.Position2D
 import it.unibo.alchemist.model.interfaces.Reaction
-import it.unibo.alchemist.model.interfaces.environments.Physics2DEnvironment
-import it.unibo.alchemist.model.interfaces.movestrategies.TargetSelectionStrategy
+import it.unibo.alchemist.model.interfaces.geometry.Vector2D
 
 /**
  * Move the pedestrian towards positions of the environment with a high concentration of the target molecule.
  *
  * @param env
  *          the environment inside which the pedestrian moves.
+ * @param reaction
+ *          the reaction which executes this action.
  * @param pedestrian
  *          the owner of this action.
  * @param targetMolecule
  *          the {@link Molecule} you want to know the concentration in the different positions of the environment.
  */
-open class FollowFlowField(
-    env: Physics2DEnvironment<Number>,
+open class FollowFlowField<P>(
+    env: Environment<Number, P>,
     reaction: Reaction<Number>,
     pedestrian: Pedestrian2D<Number>,
     targetMolecule: Molecule
-) : FlowFieldSteeringAction<Number>(
-    env,
-    reaction,
-    pedestrian,
-    targetMolecule,
-    { molecule ->
-        val currentPosition = env.getPosition(pedestrian)
-        val layer = env.getLayer(molecule).get()
-        val currentConcentration = layer.getValue(currentPosition).toDouble()
-        this.map { it to layer.getValue(it).toDouble() }
-            .filter { it.second > currentConcentration }
-            .maxBy { it.second }?.first ?: currentPosition
-    },
-    TargetSelectionStrategy {
-        with(env.getLayer(targetMolecule)) {
-            if (isEmpty || get() !is BidimensionalGaussianLayer<*>) {
-                env.origin()
-            } else {
-                with(get() as BidimensionalGaussianLayer) {
-                    env.makePosition(centerX, centerY)
-                }
+) : FlowFieldSteeringAction<P>(env, reaction, pedestrian, targetMolecule)
+    where
+        P : Position2D<P>,
+        P : Vector2D<P> {
+
+    override fun cloneAction(n: Node<Number>, r: Reaction<Number>): Action<Number> =
+        FollowFlowField(env, r, n as Pedestrian2D<Number>, targetMolecule)
+
+    override fun List<P>.selectPosition(layer: Layer<Number, P>, currentConcentration: Double): P = this
```

Comment:
This method is duplicated... can its logic get refactored out?

Reply:
It's not a duplicate. `AvoidFlowField.selectPosition` selects the position with the least concentration, whereas `FollowFlowField.selectPosition` picks the highest one. A super method taking in a boolean would need to have at least two `if/else` and I think it wouldn't be so readable.

Reply:
not really.
This portion looks completely identical to me
```
        .apply {
            if (layer is BidimensionalGaussianLayer<*>) {
                /*
                 * If the layer is Gaussian (i.e. has a center), probably the most suitable
                 * position is the one obtained by moving away from the center along the
                 * direction which connects the current position to the center.
                 */
                val center = env.makePosition(layer.centerX, layer.centerY)
                this.add(currentPosition + (currentPosition - center).resized(maxWalk()))
            }
        }
        .discardUnsuitablePositions(env, pedestrian)
        .map { it to layer.concentrationIn(it) }
```

```
        .apply {
            if (layer is BidimensionalGaussianLayer<*>) {
                /*
                 * If the layer is Gaussian (i.e. has a center), probably the most suitable
                 * position is the one obtained by moving away from the center along the
                 * direction which connects the current position to the center.
                 */
                val center = env.makePosition(layer.centerX, layer.centerY)
                this.add(currentPosition + (center - currentPosition).resized(maxWalk()))
            }
        }
        .discardUnsuitablePositions(env, pedestrian)
        .map { it to layer.concentrationIn(it) }
```
there is no need to pass booleans, it just needs to get factored out.

Reply:
There's a difference, here we add a position moving away from the center as the crow flies:

`this.add(currentPosition + (currentPosition - center).resized(maxWalk()))`

while here we add a position moving towards the center:

`this.add(currentPosition + (center - currentPosition).resized(maxWalk()))`

Maybe I could factor out the code computing the center of the layer.

Reply:
yes please

Reply:
Ok, did it.